js/src/components/tree-select-control/index.js

```diff
@@ -1,56 +1,217 @@
-// import { __, _n, sprintf } from '@wordpress/i18n';
-// import { Component, createElement } from '@wordpress/element';
-// import { debounce, escapeRegExp, identity, noop } from 'lodash';
-// import PropTypes from 'prop-types';
-// import { withFocusOutside, withSpokenMessages } from '@wordpress/components';
-// import { withInstanceId, compose } from '@wordpress/compose';
+// eslint-disable-next-line import/no-extraneous-dependencies,@woocommerce/dependency-group,@wordpress/no-unsafe-wp-apis
+import {
+	__experimentalUseFocusOutside as useFocusOutside,
+	useInstanceId,
+} from '@wordpress/compose';
 
 /**
  * Internal dependencies
  */
-import List from './list';
+import useIsEqualRefValue from '.~/hooks/useIsEqualRefValue';
 import Control from './control';
+import List from './list';
 import './index.scss';
 
-const TreeSelectControl = ( props ) => {
-	const { className, disabled, value = [], onChange = () => {} } = props;
-	const [ isExpanded, setIsExpanded ] = useState( true );
+/**
+ * The Option type Object. This is how we send the options to the selector.
+ *
+ * @typedef { Object } Option
+ * @property {string} id The unique ID for the option
+ * @property {string} name The name for the option
+ * @property {Option[]} [children] The children Option objects
+ */
+
+/**
+ * Renders a component with a searchable control, tags and a tree selector.
+ *
+ * @param {Object} props Component props.
+ * @param {string} props.id Component id
+ * @param {string} props.label Label for the component
+ * @param {string} props.placeholder Placeholder for the search control input
+ * @param {string} props.className The class name for this component
+ * @param {boolean} props.disabled Disables the component
+ * @param {Option[]} props.options Options to show in the component
+ * @param {{string}[]} props.value Selected values
+ * @param {Function} props.onChange Callback when the selector changes
+ *
+ * @return {JSX.Element|null} The component
+ */
+const TreeSelectControl = ( {
+	id,
+	label,
+	placeholder,
+	className,
+	disabled,
+	options = [],
+	value = [],
+	onChange = () => {},
+} ) => {
+	let instanceId = useInstanceId( TreeSelectControl );
+	instanceId = id ?? instanceId;
+	const [ isExpanded, setIsExpanded ] = useState( false );
+	const optionsRef = useIsEqualRefValue( options );
+	const focusOutside = useFocusOutside( () => {
+		setIsExpanded( false );
+	} );
+
+	/**
+	 * Optimizes the performance for getting the tags info
+	 *
+	 * @see getTags
+	 */
+	const optionsRepository = useMemo( () => {
+		const repository = {};
+
+		function loadOption( option ) {
+			if ( ! option.children ) {
+				repository[ option.id ] = { ...option };
+			}
+
+			option.children.forEach( ( child ) => {
+				repository[ child.id ] = { ...child };
+
+				if ( child.children ) {
+					loadOption( child );
+				}
+			} );
+		}
+
+		optionsRef.forEach( ( option ) => loadOption( option ) );
+
+		return repository;
+	}, [ optionsRef ] );
```

Comment:
> I found myself having to do composer install when switching branches. I have merged the latest develop branch code into this branch to fix that.

I guess that is because of some changes in PHP dependencies in new branches after the creation of the original branch. Thanks for merging develop 👌

Reply:
> We have the "US" parent node, but we don't have the other parent nodes "EU", "AS", "NA".

> Another round of review, found a breaking code marked with 🚧 

These 2 are related to a bug on my side. Thanks for catching it. :) 

Is fixed here https://github.com/woocommerce/google-listings-and-ads/pull/1243/commits/834bb4467e2237f84d37381ee957f0febdccb118

Reply:
Gotcha, I can now see the `repository` object stores id and option for leaf nodes only, no parent nodes. 👍 

<img width="464" alt="image" src="https://user-images.githubusercontent.com/417342/156185649-aaad54fc-543b-409d-9add-2508b60d15f2.png">